src/cdk-experimental/listbox/listbox.ts

```diff
@@ -194,360 +181,541 @@ export class CdkOption<T = unknown> implements ListKeyManagerOption, Highlightab
+    this.listKeyManager = new ActiveDescendantKeyManager(this.options)
+      .withWrap()
+      .withTypeAhead()
+      .withHomeAndEnd()
+      .withAllowedModifierKeys(['shiftKey']);
 
-  /** Sets the select's value. Required to implement ControlValueAccessor. */
-  writeValue(values: T | T[]): void {
-    if (this._options) {
-      this._setSelectionByValue(values);
+    if (this.orientation === 'vertical') {
+      this.listKeyManager.withVerticalOrientation();
+    } else {
+      this.listKeyManager.withHorizontalOrientation(this._dir?.value || 'ltr');
     }
+
+    this.listKeyManager.change
+      .pipe(takeUntil(this.destroyed))
+      .subscribe(() => this._focusActiveOption());
   }
 
-  /** Disables the select. Required to implement ControlValueAccessor. */
-  setDisabledState(isDisabled: boolean): void {
-    this.disabled = isDisabled;
+  // TODO(mmalerba): Should not depend on combobox.
+  private _updatePanelForSelection(option: CdkOption<T>) {
+    if (this._combobox) {
+      if (!this.multiple) {
+        this._combobox.updateAndClose(option.isSelected() ? option.value : []);
+      } else {
+        this._combobox.updateAndClose(this.value);
+      }
+    }
   }
 
-  /** Returns the values of the currently selected options. */
-  getSelectedValues(): T[] {
-    return this._options.filter(option => option.selected).map(option => option.value);
+  /** Update the selection mode when the 'multiple' property changes. */
+  private _updateSelectionModel() {
+    this.selectionModelSubject.next(
+      new SelectionModel(
+        this.multiple,
+        !this.multiple && this.value.length > 1 ? [] : this.value,
+        true,
+        this._compareWith,
+      ),
+    );
   }
 
-  /** Selects an option that has the corresponding given value. */
-  private _setSelectionByValue(values: T | T[]) {
-    for (const option of this._options.toArray()) {
-      this.deselect(option);
+  /** Focus the active option. */
+  private _focusActiveOption() {
+    if (!this.useActiveDescendant) {
+      this.listKeyManager.activeItem?.focus();
     }
+    this.changeDetectorRef.markForCheck();
+  }
 
-    const valuesArray = coerceArray(values);
-    for (const value of valuesArray) {
-      const correspondingOption = this._options.find((option: CdkOption<T>) => {
-        return option.value != null && this.compareWith(option.value, value);
-      });
+  /**
+   * Set the selected values.
+   * @param value The list of new selected values.
+   */
+  private _setSelection(value: T[]) {
+    this.selectionModel().setSelection(...this._coerceValue(value));
+  }
+
+  /** Update the internal value of the listbox based on the selection model. */
+  private _updateInternalValue() {
+    const selectionSet = new Set(this.selectionModel().selected);
+    // Reduce the options list to just the selected values, maintaining their order,
+    // but removing any duplicate values.
+    this._value = (this.options ?? []).reduce<T[]>((result, option) => {
```

Comment:
You can use `||` instead of nullish coalescing here since it'll produce less code.